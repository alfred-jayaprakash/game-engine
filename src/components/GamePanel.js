import React, {useState} from 'react';
import {
  Container,
  Row,
  Col,
  Card,
  CardImg,
  CardBody,
  Button,
  Input,
  ListGroup,
  ListGroupItem,
} from 'reactstrap';

const GamePanel = props => {
  const [answer, setAnswer] = useState ('');
  const [answers, setAnswers] = useState ([]);

  const onSubmit = e => {
    if (answer === '') {
      return console.log ('Answer cannot be empty');
    }
    if (answer.length > 10) {
      return console.log ('Answer length greater than 20');
    }

    if (answers.indexOf (answer) !== -1) {
      return console.log ('Word already guessed');
    }

    answers.push (answer);
    props.onAnswer (answer);
    setAnswer ('');
  };

  return (
    <div className="centered.form__box">
      <Card>
        <CardImg top width="100%" src={props.imgsrc} alt="Game Image" />
        <CardBody>
          <Container>
            <Row>
              <Col>
                <Input
                  type="answer"
                  name="answer"
                  id="userAnswer"
                  placeholder="Type your word"
                  onKeyDown={e => {
                    if (e.key === 'Enter') {
                      e.preventDefault ();
                      onSubmit (e);
                    }
                  }}
                  onChange={e => setAnswer (e.target.value)}
                  value={answer}
                />
                <Button color="success" onClick={onSubmit}>Submit</Button>
              </Col>
            </Row>
            <Row>
              <Col>
                {answers.map (answer => <Button key={answer}>{answer}</Button>)}
              </Col>
            </Row>
          </Container>
        </CardBody>
      </Card>
    </div>
  );
};

export default GamePanel;
