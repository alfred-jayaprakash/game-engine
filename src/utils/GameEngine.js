import io from 'socket.io-client';
const ENDPOINT = process.env.REACT_APP_SERVER_URL || '/';

let socket = null;

var GameEngine = (function () {
  return {
    connect: function () {
      socket = io (ENDPOINT);
      socket.on ('disconnect', () => {
        //Nothing for now
      });
    },
    join: function (username, room, callback) {
      socket.emit ('join', {username, room}, (error, data) => {
        callback (error, data);
      });
    },
    registerForRoomNotifications: function (callback) {
      socket.on ('room_data', data => {
        callback (data);
      });
    },
    sendGameStatus: function (gameData, callback) {
      socket.emit ('game_status', gameData, data => {
        callback (data);
      });
    },
    registerForGameStatus: function (callback) {
      socket.on ('game_status', data => {
        callback (data);
      });
    },
    registerForGameStateUpdates: function (callback) {
      socket.on ('state', data => {
        callback (data);
      });
    },
    disconnect: function () {
      //socket = null;
    },
  };
}) ();

export default GameEngine;
