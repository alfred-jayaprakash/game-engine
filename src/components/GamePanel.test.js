import React from 'react';
import {render, fireEvent, screen, act, waitFor} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import {shallow} from 'enzyme';
import GamePanel from './GamePanel';
import GameEngine from '../utils/GameEngine';

let onAnswer;
let container;
const imageUrl = './1.jpg';
describe ('Snapshot tests', () => {
  beforeAll (() => {
    //ARRANGE
    GameEngine.connect ();
  });

  test ('should render correctly without issues', () => {
    onAnswer = () => {};
    const component = shallow (
      <GamePanel currentImage={imageUrl} onAnswer={onAnswer} />
    );

    expect (component).toMatchSnapshot ();
  });
});

describe ('Functional tests', () => {
  beforeAll (() => {
    //ARRANGE
    GameEngine.connect ();
    onAnswer = jest.fn ();
  });

  beforeEach (() => {
    container = document.createElement ('div');
    document.body.appendChild (container);
  });

  afterEach (() => {
    document.body.removeChild (container);
    container = null;
    onAnswer.mockClear ();
  });

  test ('initial screen should load correctly', () => {
    render (
      <GamePanel currentImage={imageUrl} onAnswer={onAnswer} />,
      container
    );

    //Make sure Submit button to be present
    expect (screen.queryByText ('Submit')).toBeInTheDocument ();
  });

  test ('Empty guess should not work', () => {
    render (
      <GamePanel currentImage={imageUrl} onAnswer={onAnswer} />,
      container
    );

    //Make sure Submit button to be present
    act (() => {
      let guessInput = screen.queryByPlaceholderText ('Type your word');
      let submitButton = screen.queryByText ('Submit');
      fireEvent.click (submitButton);
    });

    expect (
      screen.getByText ('Guess cannot be empty', {exact: false}) //Error message should appear
    ).toBeInTheDocument ();
    expect (onAnswer).not.toHaveBeenCalled (); //onAnswer should not be called
  });

  test ('1-letter guess should not work', () => {
    const {getByPlaceholderText} = render (
      <GamePanel currentImage={imageUrl} onAnswer={onAnswer} />,
      container
    );

    let guessInput = getByPlaceholderText ('Type your word');
    userEvent.type (guessInput, '1');
    let submitButton = screen.queryByText ('Submit');
    fireEvent.click (submitButton);

    expect (
      screen.getByText ('Guess is too short', {exact: false}) //Error message should appear
    ).toBeInTheDocument ();
    expect (onAnswer).not.toHaveBeenCalled (); //onAnswer should not be called
  });

  test ('Super long guesses should not work', () => {
    const {getByPlaceholderText} = render (
      <GamePanel currentImage={imageUrl} onAnswer={onAnswer} />,
      container
    );

    let guessInput = getByPlaceholderText ('Type your word');
    userEvent.type (guessInput, 'abcdefghijklmnopqrstu');
    let submitButton = screen.queryByText ('Submit');
    fireEvent.click (submitButton);

    expect (
      screen.getByText ('Guess cannot be longer than 20', {exact: false}) //Error message should appear
    ).toBeInTheDocument ();
    expect (onAnswer).not.toHaveBeenCalled (); //onAnswer should not be called
  });

  test ('new guess should work', () => {
    const {getByPlaceholderText} = render (
      <GamePanel currentImage={imageUrl} onAnswer={onAnswer} />,
      container
    );

    let guessInput = getByPlaceholderText ('Type your word');
    userEvent.type (guessInput, 'one');
    let submitButton = screen.queryByText ('Submit');
    fireEvent.click (submitButton);

    expect (
      screen.getByText ('one', {exact: false}) //Word should appear
    ).toBeInTheDocument ();
    expect (onAnswer).toHaveBeenCalled (); //onAnswer should be called
    expect (onAnswer.mock.calls.length).toEqual (1); //Called exactly once
    expect (onAnswer.mock.calls[0][0]).toBe ('one'); //First time should be 'one'
  });

  test ('duplicate guess should not work', () => {
    const {getByPlaceholderText} = render (
      <GamePanel currentImage={imageUrl} onAnswer={onAnswer} />,
      container
    );

    let guessInput = getByPlaceholderText ('Type your word');
    let submitButton = screen.queryByText ('Submit');

    userEvent.type (guessInput, 'one');
    fireEvent.click (submitButton);

    userEvent.type (guessInput, 'one');
    fireEvent.click (submitButton);

    expect (
      screen.getByText ('Word already guessed', {exact: false}) //Error message should appear
    ).toBeInTheDocument ();
    expect (onAnswer).toHaveBeenCalled (); //onAnswer should be called
    expect (onAnswer.mock.calls.length).toEqual (1); //Called exactly twice
  });

  test ('duplicate guess in different case should not work', () => {
    const {getByPlaceholderText} = render (
      <GamePanel currentImage={imageUrl} onAnswer={onAnswer} />,
      container
    );

    let guessInput = getByPlaceholderText ('Type your word');
    let submitButton = screen.queryByText ('Submit');

    userEvent.type (guessInput, 'OnE');
    fireEvent.click (submitButton);

    userEvent.type (guessInput, 'oNe');
    fireEvent.click (submitButton);

    expect (
      screen.getByText ('Word already guessed', {exact: false}) //Error message should appear
    ).toBeInTheDocument ();
    expect (onAnswer).toHaveBeenCalled (); //onAnswer should be called
    expect (onAnswer.mock.calls.length).toEqual (1); //Called exactly twice
  });

  test ('unique second guess should work', () => {
    const {getByPlaceholderText} = render (
      <GamePanel currentImage={imageUrl} onAnswer={onAnswer} />,
      container
    );

    let guessInput = getByPlaceholderText ('Type your word');
    let submitButton = screen.queryByText ('Submit');

    userEvent.type (guessInput, 'one');
    fireEvent.click (submitButton);

    userEvent.type (guessInput, 'two');
    fireEvent.click (submitButton);

    expect (onAnswer).toHaveBeenCalled (); //onAnswer should be called
    expect (onAnswer.mock.calls.length).toEqual (2); //Called exactly twice
    expect (onAnswer.mock.calls[0][0]).toBe ('one'); //First time should be 'one'
    expect (onAnswer.mock.calls[1][0]).toBe ('two'); //Second time should be 'two'
  });

  test ('new image should clear previous answers', async () => {
    const {queryByText, rerender, getByPlaceholderText} = render (
      <GamePanel currentImage={imageUrl} onAnswer={onAnswer} />,
      container
    );

    let guessInput = getByPlaceholderText ('Type your word');
    let submitButton = screen.queryByText ('Submit');

    userEvent.type (guessInput, 'one');
    fireEvent.click (submitButton);

    userEvent.type (guessInput, 'two');
    fireEvent.click (submitButton);

    userEvent.type (guessInput, 'three');
    fireEvent.click (submitButton);

    expect (queryByText ('one')).toBeInTheDocument ();
    expect (queryByText ('two')).toBeInTheDocument ();
    expect (queryByText ('three')).toBeInTheDocument ();

    // re-render with different image
    let newImageUrl = './2.jpg';
    rerender (<GamePanel currentImage={newImageUrl} onAnswer={onAnswer} />);
    expect (queryByText ('one')).not.toBeInTheDocument ();

    // guessInput = getByPlaceholderText ('Type your word');
    // submitButton = screen.queryByText ('Submit');

    // userEvent.type (guessInput, 'north');
    // fireEvent.click (submitButton);

    // userEvent.type (guessInput, 'south');
    // fireEvent.click (submitButton);

    // userEvent.type (guessInput, 'west');
    // fireEvent.click (submitButton);
    // // Again render with a different image
    // newImageUrl = './3.jpg';
    // rerender (<GamePanel currentImage={newImageUrl} onAnswer={onAnswer} />);
    // expect (queryByText ('north')).not.toBeInTheDocument ();
  });
});
