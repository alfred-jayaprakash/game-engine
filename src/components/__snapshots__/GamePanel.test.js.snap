// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot tests should render correctly without issues 1`] = `
<div>
  <Row
    tag="div"
    widths={
      Array [
        "xs",
        "sm",
        "md",
        "lg",
        "xl",
      ]
    }
  >
    <Col
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <Card
        tag="div"
      >
        <CardImg
          alt="Game Image"
          height="256"
          src="./1.jpg"
          tag="img"
          top={true}
          width="256"
        />
        <CardBody
          tag="div"
        >
          <TimerPanel
            currentImage="./1.jpg"
          />
          <Form
            tag="form"
          >
            <FormGroup
              tag="div"
            >
              <Input
                className="form-control form-control-lg"
                id="userAnswer"
                invalid={false}
                name="answer"
                onChange={[Function]}
                onKeyDown={[Function]}
                placeholder="Type your word"
                type="text"
                value=""
              />
            </FormGroup>
            <FormGroup
              tag="div"
            >
              <Button
                className="btn-lg btn-block"
                color="info"
                id="guessBtn"
                onClick={[Function]}
                tag="button"
                type="button"
              >
                Submit
              </Button>
            </FormGroup>
          </Form>
        </CardBody>
      </Card>
    </Col>
  </Row>
  <Row
    tag="div"
    widths={
      Array [
        "xs",
        "sm",
        "md",
        "lg",
        "xl",
      ]
    }
  >
    <Col
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <ListGroup
        horizontal={false}
        tag="ul"
      />
    </Col>
  </Row>
</div>
`;
